### ISSUE TITLE
[Mono] Consider moving Godot data structures to a new namespace to avoid class name collisions

### ISSUE THREAD

**Issue by Jcozman** [godotengine/godot#20610]
**Godot version:**

3.0.6

**OS/device including version:**

Irrelevant 

**Issue description:**

Currently, Godot's C# implementation of its `Dictionary` class collides with the dictionary class in `System.Collections.Generic`. Often, both the `Godot` and `Systems.Collections.Generic` namespaces must be used, which requires more specificity in the code when declaring dictionary types.

Perhaps Godot data structures could be moved to a namespace like `Godot.Collections` to avoid such collisions.

**Steps to reproduce:**
Irrelevant

**Minimal reproduction project:**

Irrelevant

**Example error**
```
error CS0104: 'Dictionary<,>' is an ambiguous reference between 'Godot.Dictionary<TKey, TValue>' and 'System.Collections.Generic.Dictionary<TKey, TValue>'
```

**Comment by hpvb** [comment]
I would suggest implementing the rest of the System.Collections.Generic datastructures in the C# bindings also. This way there is no reason to import System.Collections.Generic at all. There aren't all that many and we can add the serialization/deserialization stuff to all of them.

**Comment by redxdev** [comment]
Aliases are a pain to use in the meantime, by the way. C# doesn't allow for generic namespace aliases and instead requires you to have a separate alias for each generic type:

    using StringDictionary = System.Collections.Generic.Dictionary<string, string>; // oof

Not pretty if you need to use different generics with your dictionaries.

> This way there is no reason to import System.Collections.Generic at all.

Are you suggesting that `System.Collections.Generic` be reimplemented and the replacements be recommended over it? That's really not a great idea, LINQ depends on `System.Collections.Generic.IEnumerable<T>`, pretty much every C# library in existence uses those classes, etc. Why reinvent the wheel?

**Comment by exts** [comment]
@neikeq I think that's the best solution personally, that way opt in is easier. I'm also not opposed to TMM's suggestion to just implement all of c#'s collections w/ the serialize changes. Regardless of what you do it's early enough that not everyone will be using `Godot.Dictionary` to make said change.

**Comment by PJB3005** [comment]
@hpvb Performance, for one. If I need to store a list purely in my C# code, I absolutely will use `System.Collections.Generic.List<T>` because it will perform much better than going through Godot. 

Also, can you store non-`Godot.Object`/marshall-able things in the Godot collections?

**Comment by NathanWarden** [comment]
I would also vote against re-implementing all of System.Collections.Generic unless there's a *very* compelling reason to do so.

I like @neikeq 's idea to move Array and Dictionary into `Godot.Collections`. Even in large projects this shouldn't take but a few minutes (if that) for people to add to the top of their cs files.

**Comment by nmearl** [comment]
This is also true for arrays; there's now a collision between `Godot.Array` and `System.Array`. By default, both `System` and `Godot` are imported. This is wreaking havoc on my C# project since nominally, an array object initialized via e.g. `float[]` is a `System.Array`. An array of a node class will complain that `Godot.Array` cannot be cast to a `System.Array`.

**Comment by exts** [comment]
Is it possible to get this in 3.0.7 @neikeq or if it's as simple as adding a namespace to Array.cs and Dictionary.cs I'll make a PR.

**Comment by aaronfranke** [comment]
I would appreciate this fix. Currently I alias my dictionary types since they're just too long to type. If I didn't need to prepend "System.Collections.Generic" then that would be nice.

    using IntDict = System.Collections.Generic.Dictionary<int, int>;

**Comment by exts** [comment]
@aaronfranke https://github.com/godotengine/godot/pull/21423 should be available with 3.0.7
